hedera-node/docs/design/services/token-service/airdrops/token-cancel-airdrop.md

```diff
@@ -0,0 +1,140 @@
+}
+
+/**
+ * A unique, composite, identifier for a pending airdrop.
+ *
+ * Each pending airdrop SHALL be uniquely identified by a PendingAirdropId.
+ * A PendingAirdropId SHALL be recorded when created and MUST be provided in any transaction
+ * that would modify that pending airdrop (such as a `claimAirdrop` or `cancelAirdrop`).
+ */
+message PendingAirdropId {
+    /**
+     * A sending account.<br/>
+     * This is the account that initiated, and SHALL fund, this pending airdrop.<br/>
+     * This field is REQUIRED.
+     */
+    AccountID sender_id = 1;
+
+    /**
+     * A receiving account.<br/>
+     * This is the ID of the account that SHALL receive the airdrop.<br/>
+     * This field is REQUIRED.
+     */
+    AccountID receiver_id = 2;
+
+    oneof token_reference {
+        /**
+         * A token ID.<br/>
+         * This is the type of token for a fungible/common token airdrop.<br/>
+         * This field is REQUIRED for a fungible/common token and MUST NOT be used for a
+         * non-fungible/unique token.
+         */
+        TokenID fungible_token_type = 3;
+
+        /**
+         * The id of a single NFT, consisting of a Token ID and serial number.<br/>
+         * This is the type of token for a non-fungible/unique token airdrop.<br/>
+         * This field is REQUIRED for a non-fungible/unique token and MUST NOT be used for a
+         * fungible/common token.
+         */
+        NftID non_fungible_token = 4;
+    }
+}
+```
+
+Add new RPC to `TokenService` :
+
+```protobuf
+service TokenService {
+
+//    ...
+
+    /**
+     * Cancel one or more pending airdrops.<br/>
+     * This transaction MUST be signed by _each_ account *sending* an airdrop to be canceled.
+     */
+    rpc cancelAirdrop (Transaction) returns (TransactionResponse);
+}
+```
+
+### Fees
+
+A sender may cancel pending transfers for a low nominal fee using the new `TokenCancelAirdrop` transaction. The fee should be proportional to the number of airdrops being canceled in the transaction.
+
+An update into the `feeSchedule` file would be needed to specify that.
+
+### Services updates
+
+- Update `TokenServiceDefinition` class to include the new RPC method definition for cancelling airdrops
+- Implement new `TokenCancelAirdropHandler` class which should be invoked when the gRPC server handles `TokenCancelAirdrop` transactions. The class should be responsible for:
+    - Pure checks: validation logic based only on the transaction body itself in order to verify if the transaction is valid one
+        - Verify that the pending airdrops list contains between 1 and 10 entries, inclusive
+        - Verify that the pending airdrops list does not have any duplicate entries
+    - Pre-handle:
+        - The transaction must be signed by the account referenced by a `sender_id` for each entry in the pending airdrops list
+        - Confirm that for the given pending airdrops ids in the transaction there are corresponding pending transfers existing in state
```

Comment:
No, this must be checked in `handle()`. The method `preHandle()` must not check state except if it is related to signature verification.

Reply:
Updated.